src/main/java/org/ambraproject/rhino/view/article/RelationshipViewFactory.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2019 Public Library of Science
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package org.ambraproject.rhino.view.article;
+
+import java.time.LocalDate;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.google.common.collect.ImmutableMap;
+
+import org.ambraproject.rhino.identity.ArticleIdentifier;
+import org.ambraproject.rhino.model.Article;
+import org.ambraproject.rhino.model.ArticleIngestion;
+import org.ambraproject.rhino.model.ArticleRelationship;
+import org.ambraproject.rhino.model.ArticleRevision;
+import org.ambraproject.rhino.service.ArticleCrudService;
+import org.ambraproject.rhino.view.journal.JournalOutputView;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class RelationshipViewFactory {
+  @Autowired
+  private ArticleCrudService articleCrudService;
+
+  public static Map<String, String> invertedTypes = ImmutableMap.<String, String>builder()
+    .put("commentary", "commentary-article")
+    .put("commentary-article", "commentary")
+    .put("companion", "companion")
+    .put("corrected-article", "correction-forward")
+    .put("correction-forward", "corrected-article")
+    .put("retracted-article", "retraction-forward")
+    .put("retraction-forward", "retracted-article")
+    .put("object-of-concern", "concern-forward")
+    .put("concern-forward", "object-of-concern")
+    .put("updated-article", "update-forward")
+    .put("update-forward", "updated-article")
+    .build();
+
+  /* 
+   * Our corpus contains some dirty data: make it canonical.
+   */
+  private static String getCanonicalType(String type) {
+    if (type.equals("corrrection-forward")) {
```

Comment:
Do we need to put some sort of validation in place to ensure that we aren't adding more crap like this into the DB? The XML Groomer might handle this, but if not, maybe we should create a ticket.

Reply:
Good idea, although that would require further hard-coding of a list of acceptable article relation types.

Reply:
I was thinking an FK and a presence constraint in the db, but I here ya.

Reply:
But the XML groomer doesn't connect to the db. I guess we could have some sort of rhino API to retrieve the list of correct article relation types.

Reply:
sorry, I mean, an FK and presence constraint that's enforced in rhino, that would get triggered at ingestion-time I guess. 

I'm a little nervous about depending on XMLGroomer for validation of data in Rhino, but it might already be doing this job pretty well.

Reply:
It probably would be better to have some sort rhino check. I'm not sure how XML groomer is used - it is perhaps used directly by Apex? It also provides helpful error messages.